contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -2867,6 +2879,377 @@ pltsql_detect_numeric_overflow(int weight, int dscale, int first_block, int nume
+					 */
+					switch (len)
+					{
+						case 1:
+						case 2:
+							/*
+							 * For example if the input is "1 am"
+							 * will be converted to "1:00:00 am".
+							 */
+							if (i == nf - 1 || ftype[i+1] != DTK_STRING)
+								TIME_IN_ERROR();
+							if (pg_strcasecmp(field[i+1], "am") == 0 || pg_strcasecmp(field[i+1], "pm") == 0)
+							{
+								res = makeStringInfo();
+								appendStringInfo(res, "%s%s", field[i], ":00:00");
+								field[i] = res->data;
+								ftype[i] = DTK_TIME;
+							}
+							else
+								TIME_IN_ERROR();
+							break;
+
+						case 4:
+							/*
+							 * For example if input is "2000" will be converted
+							 * to "2000-01-01".
+							 */
+							res = makeStringInfo();
+							appendStringInfo(res, "%s%s", field[i], "-01-01");
+							field[i] = res->data;
+							ftype[i] = DTK_DATE;
+							break;
+
+						case 6:
+						case 8:
+							/*
+							 * For example if input is "201213" will be converted
+							 * to "2000-12-13", and input "20001118" will be
+							 * converted to "2000-11-18"
+							 */
+							res = makeStringInfo();
+							for (int k = 0; k < len; k++)
+							{
+								appendStringInfo(res, "%c", temp_field[k]);
+								if ((len == 6 && ((k + 1) % 2 == 0 && k < 5)) ||
+										(len == 8 && (k == 3 || k == 5)))
+									appendStringInfo(res, "%c", '-');
+							}
+							field[i] = res->data;	
+							ftype[i] = DTK_DATE;
+							break;
+
+						default:
+							TIME_IN_ERROR();
+					}
+					break;
+
+				case DTK_DATE:
+					/*
+					 * If the input is of format '<DATE>T<TIME>, then the
+					 * <DATE> should be of format "YYYY-MM-DD"
+					 */
+					if (i == 0 && nf > 1 && field[1] && pg_strcasecmp(field[1], "t") == 0)
+					{
+						pattern = "^[1-9][0-9]{3}-[0-9]?[0-9]-[0-9]?[0-9]$";
+						if (regcomp(&time_regex, pattern, REG_EXTENDED) != 0)
+							ereport(ERROR,
+										(errcode(ERRCODE_INTERNAL_ERROR),
+										errmsg("time format regex failed")));
+						if (regexec(&time_regex, field[i], 0, NULL, 0) != 0)
+							TIME_IN_ERROR();
+						regfree(&time_regex);
+					}
+					else
+					{
+						/*
+						* If the date of input is of format `Mon{/.-}yyyy{/.-}dd`, `Mon{/.-}dd{/.-}yyyy
+						* "DD MM YYYY", "DD YYYY MM", "MM YYYY DD", "MM DD YYYY", "YYYY DD MM",
+						* "YYYY MM DD", "DD{/.-}MM{/.-}YYYY", "DD{/.-}YYYY{/.-}DD"
+						* then these dates shouldn't be supported.
+						* Supported date formats are `Mon yyyy dd`, `Mon dd yyyy`,
+						* `mm{-/.}dd{-/.}yyyy`, `yyyy{-/.}mm{-/.}dd`
+						*/
+						for (int k = 0 ; k < 3; k++)
+						{
+							temp_field = strtok(temp_field, different_date_formats[k]);
+							if (pg_strcasecmp(field[i], temp_field) != 0)
+								break;
+						}
+						/*
+						* If text month is present in start of field like
+						* `Mon{/.-}yyyy{/.-}dd` or `Mon{/.-}dd{/.-}yyyy then
+						* an error should be thrown
+						*/
+						if (isTextMonthPresent(temp_field))
```

Comment:
how is this possible, we are putting Text month at position 2 (in 1 to 3)

Reply:
We are putting the Text Month to middle only if the input contains date parts separated like "DD MON YYYY" but earlier didn't put the "MON-DD-YYYY" the pattern which doesn't even fall under category of "MON DD YYYY". So separate handling has been done.